File: frontend/src/services/api/bot-memory.ts
/**
 * Bot memory store API module.
 * Provides per-bot persistent key-value memory CRUD operations.
 */
import { apiFetch } from './client';

export interface MemoryEntry {
  bot_id: string;
  key: string;
  value: string;
  source: 'bot' | 'manual';
  expires_at: string | null;
  updated_at: string;
  used_bytes?: number;
  max_bytes?: number;
}

export interface BotMemorySummary {
  bot_id: string;
  bot_name: string;
  entry_count: number;
  used_bytes: number;
}

export interface BotMemoryListAllResponse {
  bots: BotMemorySummary[];
  total: number;
}

export interface BotMemoryResponse {
  bot_id: string;
  entries: MemoryEntry[];
  used_bytes: number;
  max_bytes: number;
}

export interface UpsertMemoryEntryRequest {
  value: string;
  expiresAt?: string | null;
}

export const botMemoryApi = {
  /** List all bots that have memory entries. */
  listAll: () =>
    apiFetch<BotMemoryListAllResponse>('/admin/bots/memory'),

  /** Get all memory entries for a specific bot. */
  getBotMemory: (botId: string) =>
    apiFetch<BotMemoryResponse>(`/admin/bots/${botId}/memory`),

  /** Create or update a memory entry. */
  upsertEntry: (botId: string, key: string, body: UpsertMemoryEntryRequest) =>
    apiFetch<MemoryEntry>(`/admin/bots/${botId}/memory/${encodeURIComponent(key)}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    }),

  /** Delete a single memory entry. */
  deleteEntry: (botId: string, key: string) =>
    apiFetch<{ message: string }>(`/admin/bots/${botId}/memory/${encodeURIComponent(key)}`, {
      method: 'DELETE',
    }),

  /** Delete all memory entries for a bot. */
  clearAll: (botId: string) =>
    apiFetch<{ message: string }>(`/admin/bots/${botId}/memory`, {
      method: 'DELETE',
    }),
};

File: backend/app/db/bot_memory.py
"""Bot memory store CRUD operations.

Each bot maintains a key-value memory store that persists across executions.
Entries have a source (bot or manual), optional expiry, and a per-bot byte quota.
"""

import logging

from .connection import get_connection

logger = logging.getLogger(__name__)

MAX_BYTES = 65536


def get_bot_memory(bot_id: str) -> list[dict]:
    """Return all memory entries for a bot, with computed byte usage.

    Each returned entry includes ``used_bytes`` (total for the bot) and
    ``max_bytes`` (65536) so the frontend can render a usage bar.
    """
    with get_connection() as conn:
        cursor = conn.execute(
            "SELECT bot_id, key, value, source, expires_at, updated_at "
            "FROM bot_memory WHERE bot_id = ? ORDER BY updated_at DESC",
            (bot_id,),
        )
        rows = [dict(row) for row in cursor.fetchall()]

    used_bytes = sum(len(r["value"].encode("utf-8")) for r in rows)
    for row in rows:
        row["used_bytes"] = used_bytes
        row["max_bytes"] = MAX_BYTES

    return rows


def upsert_memory_entry(
    bot_id: str,
    key: str,
    value: str,
    source: str = "manual",
    expires_at: str | None = None,
) -> dict:
    """Insert or replace a memory entry. Returns the saved entry as a dict."""
    with get_connection() as conn:
        conn.execute(
            """INSERT INTO bot_memory (bot_id, key, value, source, expires_at, updated_at)
               VALUES (?, ?, ?, ?, ?, datetime('now'))
               ON CONFLICT(bot_id, key) DO UPDATE SET
                   value = excluded.value,
                   source = excluded.source,
                   expires_at = excluded.expires_at,
                   updated_at = datetime('now')""",
            (bot_id, key, value, source, expires_at),
        )
        conn.commit()
        cursor = conn.execute(
            "SELECT bot_id, key, value, source, expires_at, updated_at "
            "FROM bot_memory WHERE bot_id = ? AND key = ?",
            (bot_id, key),
        )
        row = cursor.fetchone()
        return dict(row) if row else {}


def delete_memory_entry(bot_id: str, key: str) -> bool:
    """Delete a single memory entry. Returns True if a row was deleted."""
    with get_connection() as conn:
        cursor = conn.execute(
            "DELETE FROM bot_memory WHERE bot_id = ? AND key = ?",
            (bot_id, key),
        )
        conn.commit()
        return cursor.rowcount > 0


def clear_bot_memory(bot_id: str) -> bool:
    """Delete all memory entries for a bot. Returns True if any rows were deleted."""
    with get_connection() as conn:
        cursor = conn.execute("DELETE FROM bot_memory WHERE bot_id = ?", (bot_id,))
        conn.commit()
        return cursor.rowcount > 0
